### PR TITLE
Synonyms APIs - Synonyms Sets documentation

### PR DESCRIPTION
**Pull request by carlosdelest**
Add documentation for top-level synonyms sets APIs.

There is more work to come, but I'd like an initial review for the top level APIs so I can apply it to the next APIs to document 😉 

Screenshots:

localhost_8000_guide_synonym-management-apis html

localhost_8000_guide_put-synoyms-set html

localhost_8000_guide_get-synoyms-set html

localhost_8000_guide_list-synoyms-set html

localhost_8000_guide_delete-synoyms-set html

### CODE REVIEW

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (kderusso) [inline docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:21]: I _think_ we want to lowercase `synonym(s) sets`. Also, I think we want to be consistent with pluralization throughout our documentation and refer to it as a `synonyms set`? (This comment applies throughout the PR)

```suggestion
* You can only delete a synonyms set that is not in use by any <<analysis-synonym-graph-tokenfilter,synonym graph token filters>> and <<analysis-synonym-tokenfilter,synonym token filters>> that are included in an analyzer. See <<delete-synonym-set-analyzer-requirements>> for more information.
```

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (kderusso) [inline docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:35]:
```suggestion
The `synonyms_set` identifier was not provided, or the Synonyms Set can't be deleted because it does not meet the <<delete-synonym-set-analyzer-requirements, specified requirements>>.
```

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (kderusso) [inline docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:58]: Worth linking to https://www.elastic.co/guide/en/elasticsearch/reference/current/search-analyzer.html here?

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (kderusso) [inline docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:61]: Not 100% sure on this wording, but it seems a bit more concise? 

```suggestion
If any analyzers cannot be loaded, the index becomes available and the cluster status becomes <<red-yellow-cluster-status,red or yellow>> as index shards are not available.
```

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (kderusso) [inline docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:46]: You probably want to add some tests here, which will include some setup (to create the index that will be deleted). [These instructions](https://github.com/elastic/elasticsearch/tree/main/docs#test-code-snippets) provide a good getting started point.

File: docs/reference/synonyms/apis/put-synonyms-set.asciidoc
Comment (kderusso) [inline docs/reference/synonyms/apis/put-synonyms-set.asciidoc:28]: Link to the top level APIs page here?

**Review by kderusso — COMMENTED** [review]
Nice work so far! Only a couple of comments.

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:21]: I'm happy to change Synonyms Sets to lowercase. I was taking Search Applications example for this one - I'd like to have guidance from the docs team as I'm not sure of what should be the correct form 👍 

I've fixed the plurals, thanks for catching it!

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:58]: it can be a bit redundant as it's linked in the previous line 🤔 . What does the docs team think about repeating links on adjacent paragraphs, should we do that?

Comment (kderusso) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:58]: Hmm that's a fair point 👍

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:61]: I like that! Did a minor correction and updated it 💯

File: docs/reference/synonyms/apis/list-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/list-synonyms-set.asciidoc:13]: 💯

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (kderusso) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:21]: The general guidance is that product names should be Upper Cased and "things" should be lower cased. Search Applications is a weird case because it refers to the branded product name (Search Applications) but at the end of the day it's a search application. 

Now that I've confused you more I'll cede to the docs team 😂

File: docs/reference/synonyms/apis/put-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/put-synonyms-set.asciidoc:28]: Nice!

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/synonyms-apis.asciidoc:12]: May be we should explain what is a synonyms set with some introduction, with something like this:

"In the past there were only two ways to manage synonyms: providing synonyms inline while defining an analyzer or providing synonyms in a file.  Both these methods have their drawbacks, the former could lead to the out-blown mappings, while the later requires keep synonym files consistent across all nodes.

Synonym Management API is a new way to define and manage synonyms through an API where synonyms will be stored in an internal system index.  A number of synonyms can be grouped in a "synonyms set", and you can have as many sets as needed"....

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/synonyms-apis.asciidoc:16]: can be used => must be used or "can only be used"

Synonym token filter can be used  only in search analyzers ...

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:21]: I think we can say this shortly: 

You can only delete a Synonyms Set that is not in use by any index analyzer.

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:53]: I am not sure if we need all these details and the whole [[delete-synonym-set-analyzer-requirements]] section.

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:84]: This point is very important, and must be included.

Also we can shorten things: "that is being used in a token filter used by an analyzer," => "used by an analyzer"

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/synonyms-apis.asciidoc:22]: <<put-synoyms-set>> => <<put-synonyms-set>> and others

File: docs/reference/synonyms/apis/list-synonyms-set.asciidoc
Comment (chriscressman) [inline docs/reference/synonyms/apis/list-synonyms-set.asciidoc:12]:
```suggestion
This API allows you to retrieve the total number of Synonyms Sets defined.
```

File: docs/reference/synonyms/apis/put-synonyms-set.asciidoc
Comment (chriscressman) [inline docs/reference/synonyms/apis/put-synonyms-set.asciidoc:32]:
```suggestion
Contains the definition of the synonym rules for this synonyms set:
```

File: docs/reference/synonyms/apis/put-synonyms-set.asciidoc
Comment (chriscressman) [inline docs/reference/synonyms/apis/put-synonyms-set.asciidoc:122]:
```suggestion
This would be equivalent to invoking <<indices-reload-analyzers>> for all indices that use the synonyms set.
```

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (chriscressman) [inline docs/reference/synonyms/apis/synonyms-apis.asciidoc:2]:
```suggestion
== Synonyms APIs
```
Plural to match API path. Removed "management" for consistency with other CRUD APIs ("management" is implied). This change would cascade through to a few other places that I didn't mark.

File: docs/reference/synonyms/apis/get-synonyms-set.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/get-synonyms-set.asciidoc:78]: What is the reason we are skipping running these doc tests?

File: docs/reference/synonyms/apis/list-synonyms-set.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/list-synonyms-set.asciidoc:7]: "Synonyms Set" => "Synonyms Sets"

File: docs/reference/synonyms/apis/list-synonyms-set.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/list-synonyms-set.asciidoc:37]: Set => Sets

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (chriscressman) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc]: My official docs opinion :smile: : plural and lowercase

File: docs/reference/synonyms/apis/put-synonyms-set.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/put-synonyms-set.asciidoc:32]: What does it mean  "synonym rules that conform the synonyms set:"?

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc:12]: I like this, but I think this should be part of a blog post explaining the past situation vs the present one. I'm not sure about including the reasons for an API to be created, and talking about "new" might get stale in a few releases from now...

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc:16]: I'm rephrasing this as "are used as part", to leave the "can" out 👍

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:21]: Changed!

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:53]: I think it's important to explain why users can't delete a synonyms set. It took me a while to understand it, and other members of the team as well.

Maybe it's a bit too much technical detail, but I would err on the side of providing more information that justifies the decision.

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:84]: This is more direct, adding that.

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc:22]: thanks!

File: docs/reference/synonyms/apis/get-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/get-synonyms-set.asciidoc:78]: I hadn't started working on tests yet - I'll give it a go

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:46]: Nice! I'll give them a go, thanks!

File: docs/reference/synonyms/apis/put-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/put-synonyms-set.asciidoc]: Chris had a better suggestion: "Contains the definition of the synonym rules for this synonyms set"

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc]: Changing that, thanks Chris!

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc]: Thank you both! 🙇

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:35]: that looks better!

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (leemthompo) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc]: I think we can still extract the key points from @mayya-sharipova's suggestion, without "time stamping" the content. For example, we can still outline the relative merits of this new API approach, like how it avoids problems from existing approaches. :)

Comment (kderusso) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc]: I like the idea of providing a value statement here, and then saving the full in-depth description for a blog post. 👍

Comment (carlosdelest) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc]: Thanks for the feedback! How about this? It's a small rephrasing from Mayya's statement that provides the value proposition and the comparison to other methods. LMKWy'allT!

> Synonym Management API provides a way to define and manage synonyms through an API where synonyms are stored in an internal system index. Related synonyms can be grouped in a "synonyms set". As many synonyms sets can be created as you need for managing synonyms.
> 
> This provides an alternative to the other synonyms definition methods:
> - Inline synonyms in an analyzer definition. This can lead to out-blown mappings.
> - Synonyms file. It requires to keep synonyms files consistent across all nodes.

Comment (kderusso) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc]: WDYT about these tweaks?

> The synonyms management API provides a convenient way to define and manage synonyms in an internal system index. Related synonyms can be grouped in a "synonyms set". As many synonyms sets can be created as you need for managing synonyms.
> 
> This provides an alternative to the other synonyms definition methods:
> 
> Defining inline synonyms in an analyzer definition, which can lead to [mappings explosion](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-explosion.html) and impact performance 
> Manually uploading a synonyms file and managing consistency across all nodes


On a side note we should probably also update [the synonym token filter doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html) to reference the API as well

File: docs/reference/synonyms/apis/delete-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/delete-synonyms-set.asciidoc]: I'll create a separate PR for doing code changes for being consistent with the plurals - I'd prefer to keep this PR as focused on docs as possible

**Comment by carlosdelest** [comment]
@kderusso , @mayya-sharipova , I added working tests to the docs 👍

File: docs/reference/synonyms/apis/put-synonyms-set.asciidoc
Comment (mayya-sharipova) [inline docs/reference/synonyms/apis/put-synonyms-set.asciidoc:50]: by the API =>  by {es} 

We usually say "by Elasticsearch"

**Review by mayya-sharipova — APPROVED** [review]
@carlosdelest Great work as usual, LGTM! +1 to merge after adding some introduction to the Synonyms Management API doc, that also explains what is "synonyms set".

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc]: Nice! I've put all this in the blender and updated the PR. Thanks for your suggestions!

> On a side note we should probably also update [the synonym token filter doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html) to reference the API as well

For sure - that's something I want to tackle separately, as I'd like to check how to reorg the different synonyms definition methods, potentially as a separate guide. I'll ask for the doc teams help on this 👍

**Comment by carlosdelest** [comment]
@leemthompo @chriscressman @kderusso , anything else that stands out as wrong here? 

I'm mostly looking for things not to do in the following PRs for the remaining APIs 😉

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (leemthompo) [inline docs/reference/synonyms/apis/synonyms-apis.asciidoc:21]:
```suggestion
When synonyms sets are used in synonym token filters, filters are applied as part of the <<search-analyzer,search analyzer>> analysis process.
```

Just trying to make this clearer, not sure if it helps though :)

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (leemthompo) [inline docs/reference/synonyms/apis/synonyms-apis.asciidoc:23]:
```suggestion
Your search results will immediately reflect the synonyms set changes.
```

Plurals get confusing here :)

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (leemthompo) [inline docs/reference/synonyms/apis/synonyms-apis.asciidoc:25]:
```suggestion
Use the following APIs to manage synonyms sets:
```

**Review by leemthompo — APPROVED** [review]
LGTM writing wise, just a couple of teeny essentially grammatical nits.

File: docs/reference/rest-api/index.asciidoc
Comment (abdonpijpelink) [inline docs/reference/rest-api/index.asciidoc:98]: Don't forget to add it to the list too (after line 51).

File: docs/reference/synonyms/apis/put-synonyms-set.asciidoc
Comment (abdonpijpelink) [inline docs/reference/synonyms/apis/put-synonyms-set.asciidoc:2]:
```suggestion
=== Create or update synonyms set
```

Comment (abdonpijpelink) [reply docs/reference/synonyms/apis/put-synonyms-set.asciidoc:2]: The convention has been to say "create or update" instead of "put" (see https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-put-template.html for example)

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (abdonpijpelink) [inline docs/reference/synonyms/apis/synonyms-apis.asciidoc:16]:
```suggestion

- Defining inline synonyms in an analyzer definition, which impacts mapping size and can lead to performance issues.
```

Comment (abdonpijpelink) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc:16]: Without this newline the list doesn't render correctly.

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (abdonpijpelink) [inline docs/reference/synonyms/apis/synonyms-apis.asciidoc:22]:
```suggestion
You can use these APIs to dynamically update synonyms sets used at search time.
```

Comment (abdonpijpelink) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc:22]: Just a suggestion. The original sentence is hard to read.

File: docs/reference/synonyms/apis/put-synonyms-set.asciidoc
Comment (abdonpijpelink) [inline docs/reference/synonyms/apis/put-synonyms-set.asciidoc:23]: I think it would be better to have a section for query parameters and a section for the request body here. See https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-put-template.html#put-index-template-api-path-params for example.

**Review by abdonpijpelink — COMMENTED** [review]
I've left a few suggestions.

File: docs/reference/rest-api/index.asciidoc
Comment (carlosdelest) [reply docs/reference/rest-api/index.asciidoc:98]: Thanks for the catch! 💯

File: docs/reference/synonyms/apis/put-synonyms-set.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/put-synonyms-set.asciidoc:23]: Thanks for the link, changing it!

File: docs/reference/synonyms/apis/synonyms-apis.asciidoc
Comment (carlosdelest) [reply docs/reference/synonyms/apis/synonyms-apis.asciidoc:16]: Good catch, thanks! 👍

**Comment by carlosdelest** [comment]
@elasticsearchmachine run elasticsearch-ci/part-3

**Comment by carlosdelest** [comment]
I'm merging this as it already has some approvals. Happy to continue addressing feedback on https://github.com/elastic/elasticsearch/pull/98064.

Thanks y'all!

**Pull request merged**
